### PR TITLE
Make bz2 import optional for `io.fits`

### PR DESCRIPTION
**Pull request by dhomeier**
### Description
Addressing #5967: removed the unconditional `import bz2` from `io.fits` so the module itself can be imported in Python installations without the `bz2` module. For these systems only the autodetection of bzip2'ed FITS files based on file signature and of course their transparent decompression will be disabled.

Notes

- I do not have a Python interpreter built without bzip2 support; thus I only made some checks for this case by forcing the import to fail (misnaming the module).

- It should be possible to mark the bzip2-tests `xfail` or check for the appropriate exceptions, if the  `_make_bzip2_file` function for creating the test files was modified to fall back to an `os.system("bzip2 -k test0.fits")` call or the like. But that would probably fail on Windows, and on other systems we should perhaps not make assumptions on the availability of the command line `bzip2` either, so it did not seem worth the effort to me.

- Should this go under "API Changes" or "Bug Fixes"?

### CODE REVIEW

File: astropy/io/fits/file.py [hunk 1]
```diff
@@ -561,11 +578,11 @@ def _open_filename(self, filename, mode, overwrite):
         # Make certain we're back at the beginning of the file
         # BZ2File does not support seek when the file is open for writing, but
         # when opening a file for write, bz2.BZ2File always truncates anyway.
-        if not (isinstance(self._file, bz2.BZ2File) and mode == 'ostream'):
+        if not (_is_bz2file(self._file) and mode == 'ostream'):
             self._file.seek(0)
 
     @classproperty(lazy=True)
-    def _mmap_available(cls):
+    def _mmap_available(self):
```
Comment (saimn) [inline astropy/io/fits/file.py]: Why this change ?

**Comment by saimn** [comment]
I think this can wait for v4.0.1 (unless there is a chance to get it in v4.0, @bsipocz ?), and would classify this as bug fix.

File: astropy/io/fits/file.py (hunk 1 above)
Comment (dhomeier) [reply astropy/io/fits/file.py]: That function had been flagged by flake8 as
`N805 first argument of a method should be named 'self'`
but like the rest of the second commit, it is not strictly related.

Comment (saimn) [reply astropy/io/fits/file.py]: Ok, but this `classproperty` thing is like a classmethod, so it should take a class as argument.

**Comment by bsipocz** [comment]
@saimn - 4.0.1 is fine and I actually prefer (keep the critical stuff in 4.0 so I see what are the blockers). Otherwise anything in 4.0.1 gets merged before 4.0 is out I'll remilestone them and ship in that release.

**Comment by bsipocz** [comment]
However, for this one, I'm not sure it actually needs a changelog entry?

**Comment by saimn** [comment]
Not having a changelog entry is also an option, it's a rare case and not really a bug in astropy.

File: astropy/io/fits/file.py (hunk 1 above)
Comment (dhomeier) [reply astropy/io/fits/file.py]: So the `@classproperty` decorator takes precedence? Could add a `# noqa: N805` comment to silence flake then.

**Comment by dhomeier** [comment]
Just thought about adding a note to the `io.fits` docs though.

File: astropy/io/fits/file.py (hunk 1 above)
Comment (saimn) [reply astropy/io/fits/file.py]: Yes, it's a class method, not an instance one. 
What are this `N*` error numbers, I don't get this with flake8 ?

Comment (saimn) [reply astropy/io/fits/file.py]:
```
In [3]: fits.file._File._mmap_available                                                                                                                                                        
Out[3]: True
```

Comment (dhomeier) [reply astropy/io/fits/file.py]: That's what I got from the command line:
```
> flake8 astropy/io/fits/file.py                                                                                          
astropy/io/fits/file.py:134:80: E501 line too long (80 > 79 characters)
astropy/io/fits/file.py:139:80: E501 line too long (80 > 79 characters)
astropy/io/fits/file.py:140:80: E501 line too long (80 > 79 characters)
astropy/io/fits/file.py:161:80: E501 line too long (82 > 79 characters)
astropy/io/fits/file.py:330:80: E501 line too long (80 > 79 characters)
astropy/io/fits/file.py:336:80: E501 line too long (81 > 79 characters)
astropy/io/fits/file.py:343:80: E501 line too long (84 > 79 characters)
astropy/io/fits/file.py:444:80: E501 line too long (81 > 79 characters)
astropy/io/fits/file.py:585:26: N805 first argument of a method should be named 'self'
```

Comment (saimn) [reply astropy/io/fits/file.py]: It seems to come from an additional plugin: https://pypi.org/project/pep8-naming/

Comment (MSeifert04) [reply astropy/io/fits/file.py]: I agree with @saimn that this change should be undone.

Since we don't fail o N805 there's no reason for a noqa, right?

File: astropy/io/fits/hdu/hdulist.py [hunk 2]
```diff
@@ -17,13 +16,21 @@
 from .image import PrimaryHDU, ImageHDU
 from astropy.io.fits.file import _File, FILE_MODES
 from astropy.io.fits.header import _pad_length
-from astropy.io.fits.util import (_is_int, _tmp_name, fileobj_closed, ignore_sigint,
-                    _get_array_mmap, _free_space_check, fileobj_mode, isfile)
+from astropy.io.fits.util import (_is_int, _tmp_name, fileobj_closed,
+                                  ignore_sigint, _get_array_mmap,
+                                  _free_space_check, fileobj_mode, isfile)
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/hdulist.py]: Since you already changed these lines maybe this is a good opportunity to list the functions alphabetically?

File: astropy/io/fits/file.py [hunk 3]
```diff
@@ -462,6 +475,10 @@ def _try_read_compressed(self, obj_or_name, magic, mode, ext=''):
             if mode in ['update', 'append']:
                 raise OSError("update and append modes are not supported "
                               "with bzip2 files")
+            if not HAS_BZ2:
+                raise ValueError(
+                    ".bz2 format files are not supported since the Python "
+                    "interpreter does not include the bz2 module")
```
Comment (MSeifert04) [inline astropy/io/fits/file.py]: I would prefer a simpler message along the lines of: `ImportError: Could not find built-in bz2 module.`

File: astropy/io/fits/hdu/hdulist.py [hunk 4]
```diff
@@ -1279,6 +1286,10 @@ def _flush_resize(self):
             if self._file.compression == 'gzip':
                 new_file = gzip.GzipFile(name, mode='ab+')
             elif self._file.compression == 'bzip2':
+                if not HAS_BZ2:
+                    raise ValueError(
```
Comment (MSeifert04) [inline astropy/io/fits/hdu/hdulist.py]: See above.

File: astropy/io/fits/file.py (hunk 1 above)
Comment (dhomeier) [reply astropy/io/fits/file.py]: Yes, the notifications are from the pep8 plugin, but that's part of the standard flake8 installation, at least with the Fink version and Conda. But if it is just a bug in flake8 I agree it should be reverted.

File: astropy/io/fits/file.py (hunk 3 above)
Comment (dhomeier) [reply astropy/io/fits/file.py]: I've just copied the exception verbatim from `get_readable_fileobj`, which is handling decompression for all other file types. But if it is too long, I would still keep both synchronised. And there is a point for making this an `ImportError`.
I am just a bit concerned a `Could not find ...` message might make the user think something with their path setup could be wrong rather than a genuine lack in their Python installation.

Comment (dhomeier) [reply astropy/io/fits/file.py]: Maybe `ImportError: This Python does not provide the bz2 module`?
Potentially does not even have to be the built-in one, if it is still possible to get the `bz2file` package to work under newer Python versions...

Comment (MSeifert04) [reply astropy/io/fits/file.py]: Yeah, that sounds good.

Or since we only support Python 3.6+ we could even use [`ModuleNotFoundError`](https://docs.python.org/3/library/exceptions.html#ModuleNotFoundError)?

File: astropy/io/fits/file.py (hunk 1 above)
Comment (MSeifert04) [reply astropy/io/fits/file.py]: So we revert it to `self` without noqa?

**Comment by MSeifert04** [comment]
@dhomeier Could you address the open points? Or is there any point which needs further discussion?

**Comment by dhomeier** [comment]
I think there is agreement on all the open points.

File: astropy/utils/data.py [hunk 5]
```diff
@@ -280,9 +280,8 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
         except ImportError:
             for fd in close_fds:
                 fd.close()
-            raise ValueError(
-                ".bz2 format files are not supported since the Python "
-                "interpreter does not include the bz2 module")
+            raise ModuleNotFoundError(
```
Comment (MSeifert04) [inline astropy/utils/data.py]: While this seems appropriate and unifies the errors encountered when builtin modules are missing this PR is marked as bugfix so we shouldn't change already existing API (like changing the exception).

File: astropy/utils/data.py [hunk 6]
```diff
@@ -307,9 +306,8 @@ def get_readable_fileobj(name_or_obj, encoding=None, cache=False,
         except ImportError:
             for fd in close_fds:
                 fd.close()
-            raise ValueError(
-                ".xz format files are not supported since the Python "
-                "interpreter does not include the lzma module.")
+            raise ModuleNotFoundError(
```
Comment (MSeifert04) [inline astropy/utils/data.py]: here as well.

File: docs/io/fits/index.rst [hunk 7]
```diff
@@ -168,11 +168,12 @@ Working with compressed files
 The :func:`open` function will seamlessly open FITS files that have been
 compressed with gzip, bzip2 or pkzip. Note that in this context we are talking
 about a FITS file that has been compressed with one of these utilities (e.g., a
-.fits.gz file).
+.fits.gz file). Opening bzip2-compressed files requires the :mod:`bz2` module
```
Comment (MSeifert04) [inline docs/io/fits/index.rst]: This makes it sound like it's an optional dependency. I think it's reasonable to expect most users to have the module so this doesn't need to be explicitly mentioned, especially not on the main `index.rst`.

File: docs/io/fits/index.rst [hunk 8]
```diff
@@ -168,11 +168,12 @@ Working with compressed files
...
+to be available.
 
 There are some limitations when working with compressed files. For example,
 with Zip files that contain multiple compressed files, only the first file will
-be accessible. Also bzip does not support the append or update access modes.
+be accessible. Also bzip2 does not support the append or update access modes.
```
Comment (MSeifert04) [inline docs/io/fits/index.rst]: The change is reasonable. However it seems the sentence itself is outdated. When I look on the [module docs](https://docs.python.org/3/library/bz2.html#bz2.BZ2File) it seems like `bz2` supports appending and overwriting. I will add a new issue for this.

File: docs/io/fits/index.rst (hunk 7 above)
Comment (dhomeier) [reply docs/io/fits/index.rst]: Hmm, the whole point of the PR was to make it an optional dependency so users with the (rare) setup of a Python built without `bz2` support would be able to use Astropy with the remaining functionality at all. It is sitting there quite prominently for such a fringe case, but I did not find any other documentation on the automatic (de)compression functionality.

File: astropy/utils/data.py (hunk 5 above)
Comment (dhomeier) [reply astropy/utils/data.py]: OK, if the rules on API changes don't allow it, I guess it's better to live with the inconsistent exceptions for a while.

File: docs/io/fits/index.rst (hunk 8 above)
Comment (dhomeier) [reply docs/io/fits/index.rst]: That would be very helpful of course, then at least [file.py](https://github.com/astropy/astropy/blob/de2be96c0b3aa1b353b872e8e97de99b1484d1b1/astropy/io/fits/file.py#L462) and probably a couple of tests will need to be updated. Not sure if `get_readable_fileobj` already supports this. `file.py` currently also rejects appending to `gzip` files; if that still holds it should probably also be mentioned in the docs. Seems a case for a new PR.

File: docs/io/fits/index.rst (hunk 7 above)
Comment (MSeifert04) [reply docs/io/fits/index.rst]: >  is sitting there quite prominently for such a fringe case, but I did not find any other documentation on the automatic (de)compression functionality.

Yeah, my suggestion would be to just not add it to the documentation at all.

File: docs/io/fits/index.rst (hunk 8 above)
Comment (MSeifert04) [reply docs/io/fits/index.rst]: Of course that should be done seperately. I just wanted to mention it so it's not forgotten.

Comment (MSeifert04) [reply docs/io/fits/index.rst]: See #9685

File: docs/io/fits/index.rst (hunk 7 above)
Comment (dhomeier) [reply docs/io/fits/index.rst]: That would probably be the alternative, as most of the `fits` documentation sits in `index.rst`. The exception message is hopefully informative enough whenever this issue might come up in the future.

Comment (saimn) [reply docs/io/fits/index.rst]: Yes I think we don't need to mention it here, since a normal Python should come with it.

**Comment by MSeifert04** [comment]
I think this looks good now.

I'm not sure if we should squash the commits to remove the back-and-forth. What do you think @saimn ?

**Comment by saimn** [comment]
Yes, looks good, thanks @dhomeier. Squashing could be useful to clean the history, if you have time, but we can also do without.

**Comment by bsipocz** [comment]
Thanks @dhomeier!

**Comment by dhomeier** [comment]
Thanks @MSeifert04 for the review and everyone for their comments!
Now that 7795c8d is in the history, do we need a new issue to get the `get_readable_fileobj` exceptions sync'ed in a future version?

**Comment by pllim** [comment]
> do we need a new issue

@dhomeier , please go ahead and open a follow-up issue if you haven't already. Thanks!
